File: frontend/src/modules/seguimiento/services/seguimientoService.js
// src/modules/seguimiento/services/seguimientoService.js

/**
 * Servicio que centraliza todas las llamadas HTTP del módulo Seguimiento.
 */
import http from '../../../shared/services/http.js';

const seguimientoService = {
  /**
   * GET /seguimiento/tablero
   * Mapea filtros del frontend a los parámetros del backend.
   */
  tablero: (params = {}) => {
    const { docenteId, estado, busqueda, ...rest } = params;
    return http.get('/seguimiento/tablero', {
      params: {
        ...rest,
        docente: docenteId || undefined,
        estadoSeguimiento: estado || undefined,
        empresa: busqueda || undefined,
      },
    });
  },

  /**
   * GET /seguimiento/practicas
   * Endpoint desacoplado: no exige programaId, resuelve el scope del usuario
   * autenticado en el backend (ESTUDIANTE ve solo la suya, coordinadores su facultad).
   */
  practicas: ({ programaId, page, size, estado, busqueda, estadoPractica } = {}) =>
    http.get('/seguimiento/practicas', {
      params: {
        programaId: programaId || undefined,
        page,
        size,
        busqueda: busqueda || undefined,
        estadoSeguimiento: estado || undefined,
        estadoPractica: estadoPractica || undefined,
      },
    }),

  /** POST /seguimiento/{practicaId}/observaciones */
  registrarObservacion: (practicaId, { corte, ...body }) =>
    http.post(`/seguimiento/${practicaId}/observaciones`, body, { params: { corte } }),

  /** POST /seguimiento/{practicaId}/avances-tutor */
  registrarAvance: (practicaId, { corte, ...body }) =>
    http.post(`/seguimiento/${practicaId}/avances-tutor`, body, { params: { corte } }),

  /** POST /seguimiento/{practicaId}/bitacora (multipart: descripcion + archivo opcional) */
  registrarBitacora: (practicaId, { descripcion, archivo }, corte = 1) => {
    const form = new FormData();
    form.append('descripcion', descripcion);
    if (archivo) form.append('archivo', archivo);
    return http.post(`/seguimiento/${practicaId}/bitacora`, form, {
      params: { corte },
      headers: { 'Content-Type': 'multipart/form-data' },
    });
  },

  /** GET /seguimiento/bitacora/{bitacoraId}/archivo — devuelve blob para descarga */
  descargarArchivoBitacora: (bitacoraId) =>
    http.get(`/seguimiento/bitacora/${bitacoraId}/archivo`, { responseType: 'blob' }),

  /** GET /seguimiento/alertas */
  alertas: () => http.get('/seguimiento/alertas'),

  /** PATCH /vinculaciones/alertas/{id}/resolver */
  marcarAlertaLeida: (id) => http.patch(`/vinculaciones/alertas/${id}/resolver`),

  /** GET /seguimiento/{practicaId} */
  obtenerDetallePractica: (id) => http.get(`/seguimiento/${id}`),

  /** GET /seguimiento/estudiantes/{estudianteId}/practicas */
  historialEstudiante: (estudianteId) =>
    http.get(`/seguimiento/estudiantes/${estudianteId}/practicas`),

  /** GET /seguimiento/{practicaId}/bitacora */
  obtenerBitacoras: (practicaId) => http.get(`/seguimiento/${practicaId}/bitacora`),
};

export default seguimientoService;

File: src/main/java/com/avh/practicas/seguimiento/service/SeguimientoService.java
package com.avh.practicas.seguimiento.service;

import com.avh.practicas.seguimiento.dto.AvanceRequest;
import com.avh.practicas.seguimiento.dto.BitacoraRequest;
import com.avh.practicas.seguimiento.dto.ObservacionRequest;
import com.avh.practicas.seguimiento.dto.TableroResponse;
import com.avh.practicas.seguimiento.entity.AlertaSistema;
import com.avh.practicas.seguimiento.entity.AvanceTutor;
import com.avh.practicas.seguimiento.entity.BitacoraEstudiante;
import com.avh.practicas.seguimiento.entity.ObservacionDocente;
import org.springframework.core.io.Resource;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

/**
 * Interfaz que define las operaciones de seguimiento de las prácticas.
 */
public interface SeguimientoService {

    /**
     * Registra una observación académica por parte de un docente asesor.
     */
    ObservacionDocente registrarObservacion(Long practicaId, Long docenteId, Integer corte, ObservacionRequest request);

    /**
     * Edita una observación existente si el corte no está cerrado.
     */
    ObservacionDocente editarObservacion(Long id, ObservacionRequest request);

    /**
     * Registra un avance de seguimiento por parte de un tutor empresarial.
     */
    AvanceTutor registrarAvanceTutor(Long practicaId, Long tutorId, Integer corte, AvanceRequest request);

    /**
     * Registra una entrada en la bitácora de actividades por parte del estudiante.
     */
    BitacoraEstudiante registrarBitacoraEstudiante(Long practicaId, Long estudianteId, Integer corte, BitacoraRequest request);

    /**
     * Retorna el tablero de seguimiento para coordinadores y secretarias.
     */
    List<TableroResponse> obtenerTableroSeguimiento(Long programaId, String empresa, String docente, Integer corte, String estadoSeguimiento);

    /**
     * Retorna las prácticas visibles para el usuario autenticado según su scope
     * (ESTUDIANTE ve solo la suya, COORD_PRACTICA/COORD_ACADEMICA ven su facultad, ADMIN/
     * SECRETARIA ven todas), sin exigir programaId. Los filtros son opcionales y acotan
     * el resultado ya cargado.
     */
    List<TableroResponse> obtenerPracticasSeguimiento(
            String busqueda,
            Long programaId,
            String estadoSeguimiento,
            String estadoPractica);

    /**
     * Obtiene todas las observaciones asociadas a una práctica.
     */
    List<ObservacionDocente> obtenerObservacionesPorPractica(Long practicaId);

    /**
     * Obtiene todos los avances de tutor asociados a una práctica.
     */
    List<AvanceTutor> obtenerAvancesPorPractica(Long practicaId);

    /**
     * Obtiene la bitácora de actividades asociada a una práctica.
     */
    List<BitacoraEstudiante> obtenerBitacorasPorPractica(Long practicaId);

    /**
     * Obtiene el detalle unificado de una práctica.
     */
    com.avh.practicas.seguimiento.dto.PracticaDetalleResponse obtenerDetallePractica(Long practicaId);

    /**
     * Adjunta un archivo de soporte a una entrada de bitácora ya creada.
     * Guarda el archivo en disco y actualiza nombre_archivo / url_archivo en la entidad.
     */
    BitacoraEstudiante adjuntarArchivoBitacora(Long bitacoraId, MultipartFile archivo);

    /**
     * Devuelve el recurso (stream del archivo) para descarga.
     */
    Resource descargarArchivoBitacora(Long bitacoraId);

    /**
     * Devuelve el nombre original del archivo adjunto de la bitácora.
     */
    String nombreArchivoBitacora(Long bitacoraId);

    /**
     * Obtiene todas las alertas activas del sistema.
     */
    List<AlertaSistema> obtenerAlertasActivas();
}
